### PR TITLE
Add typing / drop Python < 3.8 support

### PR DESCRIPTION
**Pull request by nhairs**
This PR adds Python 3.8+ compatible type annotations (dropping support for <3.8 at the same time).

Fixes: #27 

At a high level the following changes have been made:

- Update code to use new features available from 3.8 (e.g. `f"{foo=}` notation).
- Type annotations
  - note: Because we use `from __future__ import annotations` we can use "new style" annotations (e.g. `str | list[int] | None` instead of `Optional[Union[str, List[int]]]`) which are normally only available in higher versions of python. However it does mean that these annotations cannot be resolved from their string representation (all type annotations are stored as strings which is why this works) into their concrete classes unless you are running a supported version of Python.
- Completely move to `pyrproject.toml`
- Move all development dependencies to `pyproject.toml`, make `make .venv` more portable
  - `virtualenv` isn't always available so use `-m venv .venv` which is just as effective, and not all systems have `python(2)` installed or `python3-is-python` so explicitly use `python3`
- Add some new badges to README

## Test Plan
1. Ensure that `make ready-pr` runs without error

### CODE REVIEW

File: pyproject.toml
Comment (nhairs) [inline pyproject.toml:9]: @toumorokoshi - If you want to include an email address here let me know.

File: pyproject.toml
Comment (nhairs) [inline pyproject.toml:64]: Is there any preference here? I've left it at the default 88 to match your Makefile, but to be honest I usually recommend using 100 these days (most people have much larger screens than when 80 became standard).

File: deepmerge/strategy/core.py
Comment (toumorokoshi) [inline deepmerge/strategy/core.py:28]: nit: since you already use attribute parameters above (e.g. for `NAME`), I'd just annotate `_strategies` up there as well.

File: deepmerge/strategy/core.py
Comment (toumorokoshi) [inline deepmerge/strategy/core.py:51]: I intentionally didn't use the full types so it would be pretty easy to pass this through. I still kind of prefer that we don't have to have to duplicate the annotations here and in the strategy object, but I guess you kind of have to for typing reasons?

File: pyproject.toml
Comment (toumorokoshi) [reply pyproject.toml:64]: no opinion on this, go ahead and set it to what you see fit!

File: pyproject.toml
Comment (toumorokoshi) [reply pyproject.toml:13]: I think we just dropped it while adding more versions. I could go either way - I see 3.7 is EOL now too.

I agree with being consistent though - so we should have this line match the version(s) of python tested.

File: pyproject.toml
Comment (toumorokoshi) [reply pyproject.toml:9]: go for it! its yusuke@tsutsumi.io

**Comment by toumorokoshi** [comment]
to fix the tests, I think you have to remove `setup.cfg` from this line: https://github.com/toumorokoshi/deepmerge/blob/master/Makefile#L4

**Review by toumorokoshi — APPROVED** [review]
Wow, awesome! thank you so much - this is fantastic.

LGTM overall, a couple smaller nits.

Once the tests pass happy to merge it in!

File: deepmerge/strategy/core.py
Comment (nhairs) [reply deepmerge/strategy/core.py:51]: >  I guess you kind of have to for typing reasons?
Yeah pretty much - without it it's ambiguous to both callers and the type checker for what these are. It is possible to use things like a param spec, but it's really not worth it for this use case.

File: deepmerge/extended_set.py
Comment (nhairs) [inline deepmerge/extended_set.py:22]: This was renamed to prevent conflict with the `_hash` method of the parent class.

File: deepmerge/strategy/core.py
Comment (nhairs) [reply deepmerge/strategy/core.py]: The only reason I annotate `NAME` in that section is because it's a class attribute. For instance attributes I think it makes more sense to annotate them in `__init__`.

**Comment by nhairs** [comment]
Waiting on CI job being unblocked :)

**Comment by toumorokoshi** [comment]
> Made a few other changes, including the Makefile. Now waiting on CI job being unblocked :)
> 
> There is a chance this will break a lot of code so consider doing a `rc/a` release first (if you do I can test it in my own projects).
> 
> Also you'll probably want to avoid a merge commit because I have complete disregard for git history within PRs.

Thanks! I haven't had a chance to take another pass, but I did unblock CI. I'll see if I can just give your PRs permission to run without approval.

ack on RC - I was thinking the same myself, we'll ship a beta, maybe a major version too just to mark the lack of support for py2.

No worries on Git history - I'll squash at the end ;)

**Comment by nhairs** [comment]
> I'll see if I can just give your PRs permission to run without approval.

Based on other repositories I've created PRs for I believe this is possible; though I'm not sure what the setting is as I don't use GHA myself.

File: deepmerge/strategy/core.py
Comment (toumorokoshi) [inline deepmerge/strategy/core.py:9]:
```suggestion
    from typing_extensions import TypeAlias
```
I think this is what's causing 3.8/3.9 to fail.

File: deepmerge/strategy/core.py
Comment (toumorokoshi) [reply deepmerge/strategy/core.py]: I'm not strongly opinionated enough here to block the PR - but I think having the annotations in one place is nicer, although I do see the point about being unable to differentiate between class / object attributes.

Comment (nhairs) [reply deepmerge/strategy/core.py]: I can confirm that these are treated differently by type checkers. Defining a name in the class body specifies that is either a class or instance attribute (or class only using `typing.ClassVar`), whilst names defined in __init__ are instance only attributes. 

https://mypy.readthedocs.io/en/latest/class_basics.html#instance-and-class-attributes

**Comment by toumorokoshi** [comment]
awesome, thanks! will get a beta version out soon.

**Comment by toumorokoshi** [comment]
@nhairs beta is out: https://pypi.org/project/deepmerge/2.0b0/. Would you mind testing it out?

**Comment by nhairs** [comment]
Hi @toumorokoshi,

It's been 6 months since the 2.0 beta release. Would you be able to upload a full release to PyPI?

**Comment by toumorokoshi** [comment]
done! thanks for the reminder: https://pypi.org/project/deepmerge/2.0/

**Pull request merged**
